Add `parent_id` and `lock_permissions` to `reposition_channels`

rayakame opened the pull request:
### Summary
- Move `reposition_channels` to `special_endpoints` without breaking/deprecating anything.
- Add `parent_id` and `lock_permissions`for every channel.

### Checklist
<!-- Make sure to tick all the following boxes by putting an `x` in between (like this `[x]`) -->
- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### Related issues
closes #523

syncblaze committed:
first things done

syncblaze committed:
✨ refactor(rest): Refactor reposition_channels method to accept default positions.

syncblaze committed:
✨ IMPROVE(rest): Refactor reposition_channels method for better readability and maintainability

syncblaze committed:
🔧 refactor(api): Added ChannelRepositioner with request_call signature mapping.

syncblaze committed:
🔧 refactor(special_endpoints): Update ChannelRepositioner slots and initialization parameters

syncblaze committed:
✨ feat(changes): Add `parent_id` and `lock_permissions` to `rest.reposition_channels`

rayakame commented:
This should be the first Iteration of this feature.

**Tests and docs are not done yet!**
I want to clear up some things before I start working on these things.

1. **Naming**
I'm not really sure about the names yet. `ChannelRepositioner` doesn't sound right and same for `RepositionChannelHelper` .
But i'm not creative at all and wanted to ask about your guys opinions.
2. **Implementation**
With the current implementation we wouldn't need to deprecate anything and there are no breaking changes. Only problem
being that it could be pretty confusing since you can pass channels into the `rest.reposition_channels` and into the
`ChannelRepositioner`. Normally I wouldn't say that this is important enough for a breaking change or a deprecation but we
already have breaking changes planned for the next release so might as well do that.

Let me know about your opinions and I would be happy to implement them.

davfsa commented on hikari/impl/special_endpoints.py:
```diff
@@ -204,6 +205,73 @@ async def _keep_typing(self) -> None:
+class ChannelRepositioner(special_endpoints.ChannelRepositioner):
+    __slots__: typing.Sequence[str] = ("_guild", "_request_call", "_channels")
+
+    @typing.overload
+    def __init__(self, guild: snowflakes.SnowflakeishOr[guilds.PartialGuild], request_call: _RequestCallSig): ...
+
+    @typing.overload
+    def __init__(
+        self,
+        guild: snowflakes.SnowflakeishOr[guilds.PartialGuild],
+        request_call: _RequestCallSig,
+        *,
+        positions: typing.Mapping[int, snowflakes.SnowflakeishOr[channels.GuildChannel]],
+    ): ...
+
+    def __init__(
+        self,
+        guild: snowflakes.SnowflakeishOr[guilds.PartialGuild],
+        request_call: _RequestCallSig,
+        *,
+        positions: typing.Mapping[int, snowflakes.SnowflakeishOr[channels.GuildChannel]] = {},
+    ):
```
Does this need to be a typing overload? Couldnt it just be a single definition?

rayakame replied:
yeah your prob right the user shouldn't come into contact with that in the end

davfsa commented on hikari/api/special_endpoints.py:
```diff
@@ -108,6 +124,46 @@ async def __aexit__(
+
+    @typing.overload
+    @abc.abstractmethod
+    def __init__(self, guild: snowflakes.SnowflakeishOr[guilds.PartialGuild], request_call: _RequestCallSig): ...
```
We should probably copy how GuildBuilder works when it comes to taking in a request call

rayakame replied:
by making it an attrs class?

I mean I could do that but I'm doing some conversion in the constructor and I feel like a constructor is better here idk

rayakame replied:
👀

davfsa replied:
I mean that it's left as impl detail. We do not define how init is meant to behave in the api.

But tbh, I am not sure what the best way to do it is now :P

rayakame replied:
I changed a lot about the builders, including fixing your concerns here so this conversation can be closed for now.

rayakame commented:
I will take a look in a hour or so

syncblaze committed:
🔥 refactor(special_endpoints): Remove redundant code in ChannelRepositioner class

rayakame committed:
Merge branch 'master' into feature/reposition-channels

Signed-off-by: Christopher <github.android206@passmail.net>

rayakame committed:
fix ruff errors

rayakame committed:
working on changing the builder class(es)

rayakame committed:
feat(special_endpoints): refactor ChannelRepositioner class
(feature/reposition-channels)

rayakame committed:
feat(special_endpoints): refactor ChannelRepositioner class
(feature/reposition-channels)

rayakame committed:
no overloads

rayakame committed:
no deprecation :(

rayakame committed:
add missing tests

rayakame committed:
unused overloads

rayakame committed:
some docstring things

rayakame committed:
feat(rest): Add documentation for return value in reposition_channels method. (feature/reposition-channels)

rayakame commented on hikari/api/special_endpoints.py:
```diff
@@ -114,6 +114,151 @@ async def __aexit__(
+
+        """
+
+    @abc.abstractmethod
+    def __await__(self) -> typing.Generator[typing.Any, typing.Any, typing.Any]:
+        """Reposition the channels in a guild.
+
+        !!! note
+            Only channels to be modified are required.
+
+        Raises
+        ------
+        hikari.errors.UnauthorizedError
+            If you are unauthorized to make the request (invalid/missing token).
+        hikari.errors.ForbiddenError
+            If you cannot access the channel.
+        hikari.errors.NotFoundError
+            If the channel is not found.
+        hikari.errors.RateLimitTooLongError
+            Raised in the event that a rate limit occurs that is
+            longer than `max_rate_limit` when making a request.
+        hikari.errors.InternalServerError
+            If an internal error occurs on Discord while handling the request.
```
@davfsa I would need your opinion here, this would be (in my opinion) the perfect way to document the behaviour of this class, but sadly `__await__` docstrings are not displayed in the docs. What would you do here?

rayakame replied:
How about we do this:

```py
    async def reposition(self) -> None:
        await self
```

this would allow developers to call `.reposition` instead of doing `await obj` while still not breaking the `RESTClient.reposition_channels` function.

You could either do this:
```py
await rest.reposition_channels(guild=... , positions=...)
```
or this:
```py
helper = rest.reposition_channels(guild=...)
helper.add_reposition_channel(position=... , channel=...)
await helper.reposition() # instead of doing `await helper`
```

and we can put all the documentation into the `.reposition()` function for better DX

davfsa replied:
The fact that you can await it should be documented as part of the class description, if you want to make it really explicit, that would probably be the best DX. This is mostly for consistency with all the other special endpoints, and it is not something I want to change.

In the future, I can have a look at mkdocstrings and have it document `__await__` too

rayakame committed:
Merge branch 'master' into feature/reposition-channels

Signed-off-by: Christopher <github.android206@passmail.net>

rayakame committed:
fix(api): remove unnecessary spaces in ChannelRepositioner class (feature/reposition-channels)

rayakame committed:
feat(special_endpoints): Add method to reposition channels in a guild
(issue 1)

rayakame committed:
dont ask me why i yanked this in the first place

rayakame committed:
pipeline

rayakame committed:
i need to test smth

rayakame committed:
feat(special_endpoints): Add documentation for ChannelRepositioner (#feature/reposition-channels)

rayakame committed:
feat(special_endpoints): Add documentation for ChannelRepositioner (#feature/reposition-channels)

rayakame committed:
feat(special_endpoints): Add documentation for ChannelRepositioner (#feature/reposition-channels)

rayakame committed:
feat(special_endpoints): Add documentation for ChannelRepositioner (#feature/reposition-channels)

rayakame committed:
feat(special_endpoints): Add documentation for ChannelRepositioner (#feature/reposition-channels)

rayakame committed:
Merge branch 'master' into feature/reposition-channels

rayakame committed:
added last missing docstring

rayakame committed:
added last missing docstring

rayakame committed:
Merge branch 'master' into feature/reposition-channels

rayakame committed:
Merge branch 'master' into feature/reposition-channels

Signed-off-by: Christopher <github.android206@passmail.net>

rayakame committed:
feat(rest): Add reason parameter to ChannelRepositioner

Added a reason parameter to the ChannelRepositioner class in order to provide
an explanation that will be recorded in the audit logs. (issue: feature/reposition-channels)

davfsa reviewed (changes requested):
Only small things. Thanks for working on this!

rayakame committed:
Update hikari/impl/special_endpoints.py

Co-authored-by: davfsa <davfsa@gmail.com>
Signed-off-by: Christopher <github.android206@passmail.net>

rayakame committed:
Merge branch 'master' into feature/reposition-channels

rayakame committed:
refactor(special_endpoints): Change RepositionChannelHelper to RepositionedChannel in special_endpoints

Changed `RepositionChannelHelper` class to `RepositionedChannel` in special_endpoints.py
and api/special_endpoints.py files. (issue #feature/reposition-channels)

rayakame commented on hikari/api/rest.py:
```diff
@@ -5582,23 +5582,19 @@ async def reposition_channels(
         positions
             A mapping of of the object or the ID of an existing channel to
             the new position, relative to their parent category, if any.
+
+            !!! note
+                Instead of using the `positions` parameter, you should make
+                use of the returned [`hikari.api.special_endpoints.ChannelRepositioner`][].
```
I didn't intend on putting that comment there because we intend on deprecating the old way of doing this, but rather to inform people that you can/should use the `ChannelRepositioner` because otherwise you cannot use the `parent_id` and `lock_permissions`.

I think if we do not talk about it somewhere most users wouldn't even notice that you can use the `ChannelRepositioner`.

rayakame committed:
feat(rest): add documentation for errors in reposition_channels (feature/reposition-channels)

rayakame committed:
feat(api): add note to reposition channels endpoint (feature/reposition-channels)

rayakame committed:
Merge branch 'master' into feature/reposition-channels

rayakame committed:
nox pipeline

rayakame committed:
change documentation

rayakame committed:
Merge branch 'master' into feature/reposition-channels

rayakame committed:
Merge branch 'master' into feature/reposition-channels

rayakame committed:
refactor(special_endpoints): Fix typo in special endpoints documentation. (feature/reposition-channels)

rayakame committed:
refactor(rest): fix reposition_channels method to handle undefined positions properly (feature/reposition-channels)

rayakame committed:
Merge branch 'master' into feature/reposition-channels
